// docx-templates/lib/mainBrowser.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _timm = require('timm');

var _zip = require('./zip');

var _xml = require('./xml');

var _preprocessTemplate = require('./preprocessTemplate');

var _preprocessTemplate2 = _interopRequireDefault(_preprocessTemplate);

var _processTemplate = require('./processTemplate');

var _reportUtils = require('./reportUtils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable no-param-reassign, no-console */

var DEBUG = process.env.DEBUG_DOCX_TEMPLATES;
var DEFAULT_CMD_DELIMITER = '+++';
var DEFAULT_LITERAL_XML_DELIMITER = '||';

var log = DEBUG ? require('./debug').mainStory : null;

// ==========================================
// Main
// ==========================================
var createReport = function () {
  var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(options) {
    var template, data, queryVars, replaceImages, _probe, templatePath, literalXmlDelimiter, createOptions, xmlOptions, zip, templateXml, tic, parseResult, jsTemplate, tac, queryResult, query, finalTemplate, result, report1, images1, reportXml, numImages, files, images, i, filePath, raw, js0, js, _ref3, report2, images2, xml, segments, documentComponent, contentTypesPath, contentTypesXml, contentTypes, ensureContentType, finalContentTypesXml, mediaPath, imgNames, _i, imgName, imgPath, imgData, output;

    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            DEBUG && log.debug('Report options:', { attach: options });
            template = options.template, data = options.data, queryVars = options.queryVars, replaceImages = options.replaceImages, _probe = options._probe;
            templatePath = 'word';
            literalXmlDelimiter = options.literalXmlDelimiter || DEFAULT_LITERAL_XML_DELIMITER;
            createOptions = {
              cmdDelimiter: options.cmdDelimiter || DEFAULT_CMD_DELIMITER,
              literalXmlDelimiter: literalXmlDelimiter,
              processLineBreaks: options.processLineBreaks != null ? options.processLineBreaks : true,
              noSandbox: options.noSandbox || false,
              additionalJsContext: options.additionalJsContext || {}
            };
            xmlOptions = { literalXmlDelimiter: literalXmlDelimiter };

            // ---------------------------------------------------------
            // Unzip
            // ---------------------------------------------------------

            DEBUG && log.debug('Unzipping...');
            _context2.next = 9;
            return (0, _zip.zipLoad)(template);

          case 9:
            zip = _context2.sent;


            // ---------------------------------------------------------
            // Read the 'document.xml' file (the template) and parse it
            // ---------------------------------------------------------
            DEBUG && log.debug('Reading template...');
            _context2.next = 13;
            return (0, _zip.zipGetText)(zip, templatePath + '/document.xml');

          case 13:
            templateXml = _context2.sent;

            DEBUG && log.debug('Template file length: ' + templateXml.length);
            DEBUG && log.debug('Parsing XML...');
            tic = new Date().getTime();
            _context2.next = 19;
            return (0, _xml.parseXml)(templateXml);

          case 19:
            parseResult = _context2.sent;
            jsTemplate = parseResult;
            tac = new Date().getTime();

            DEBUG && log.debug('File parsed in ' + (tac - tic) + ' ms', {
              attach: jsTemplate,
              attachLevel: 'trace'
            });

            // ---------------------------------------------------------
            // Fetch the data that will fill in the template
            // ---------------------------------------------------------
            queryResult = null;

            if (!(typeof data === 'function')) {
              _context2.next = 35;
              break;
            }

            DEBUG && log.debug('Looking for the query in the template...');
            _context2.next = 28;
            return (0, _processTemplate.extractQuery)(jsTemplate, createOptions);

          case 28:
            query = _context2.sent;

            DEBUG && log.debug('Query: ' + (query || 'no query found'));
            _context2.next = 32;
            return data(query, queryVars);

          case 32:
            queryResult = _context2.sent;
            _context2.next = 36;
            break;

          case 35:
            queryResult = data;

          case 36:

            // ---------------------------------------------------------
            // Process document.xml:
            // - Generate the report
            // - Build output XML and write it to disk
            // - Images
            // ---------------------------------------------------------
            // DEBUG && log.debug('Before preprocessing...', {
            //   attach: jsTemplate,
            //   attachLevel: 'debug',
            //   ignoreKeys: ['_parent', '_fTextNode', '_attrs'],
            // });
            finalTemplate = (0, _preprocessTemplate2.default)(jsTemplate, createOptions);

            // DEBUG &&
            //   log.debug('Generating report...', {
            //     attach: finalTemplate,
            //     attachLevel: 'debug',
            //     ignoreKeys: ['_parent', '_fTextNode', '_attrs'],
            //   });

            _context2.next = 39;
            return (0, _processTemplate.produceJsReport)(queryResult, finalTemplate, createOptions);

          case 39:
            result = _context2.sent;
            report1 = result.report, images1 = result.images;

            if (!(_probe === 'JS')) {
              _context2.next = 43;
              break;
            }

            return _context2.abrupt('return', report1);

          case 43:

            // DEBUG &&
            //   log.debug('Report', {
            //     attach: report,
            //     attachLevel: 'debug',
            //     ignoreKeys: ['_parent', '_fTextNode', '_attrs'],
            //   });
            DEBUG && log.debug('Converting report to XML...');
            reportXml = (0, _xml.buildXml)(report1, xmlOptions);

            if (!(_probe === 'XML')) {
              _context2.next = 47;
              break;
            }

            return _context2.abrupt('return', reportXml);

          case 47:
            DEBUG && log.debug('Writing report...');
            (0, _zip.zipSetText)(zip, templatePath + '/document.xml', reportXml);

            numImages = (0, _keys2.default)(images1).length;

            processImages(images1, 'document.xml', zip, templatePath);

            // ---------------------------------------------------------
            // Process all other XML files (they may contain headers, etc.)
            // ---------------------------------------------------------
            files = [];

            zip.forEach(function () {
              var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(filePath) {
                var regex;
                return _regenerator2.default.wrap(function _callee$(_context) {
                  while (1) {
                    switch (_context.prev = _context.next) {
                      case 0:
                        regex = new RegExp(templatePath + '\\/[^\\/]+\\.xml');

                        if (regex.test(filePath) && filePath !== templatePath + '/document.xml') {
                          files.push(filePath);
                        }

                      case 2:
                      case 'end':
                        return _context.stop();
                    }
                  }
                }, _callee, undefined);
              }));

              return function (_x2) {
                return _ref2.apply(this, arguments);
              };
            }());

            images = images1;
            i = 0;

          case 55:
            if (!(i < files.length)) {
              _context2.next = 80;
              break;
            }

            filePath = files[i];

            DEBUG && log.info('Processing ' + filePath + '...');
            _context2.next = 60;
            return (0, _zip.zipGetText)(zip, filePath);

          case 60:
            raw = _context2.sent;
            _context2.next = 63;
            return (0, _xml.parseXml)(raw);

          case 63:
            js0 = _context2.sent;
            js = (0, _preprocessTemplate2.default)(js0, createOptions);
            _context2.next = 67;
            return (0, _processTemplate.produceJsReport)(queryResult, js, createOptions);

          case 67:
            _ref3 = _context2.sent;
            report2 = _ref3.report;
            images2 = _ref3.images;

            images = (0, _timm.merge)(images, images2);
            xml = (0, _xml.buildXml)(report2, xmlOptions);

            (0, _zip.zipSetText)(zip, filePath, xml);

            numImages += (0, _keys2.default)(images2).length;
            segments = filePath.split('/');
            documentComponent = segments[segments.length - 1];

            processImages(images2, documentComponent, zip, templatePath);

          case 77:
            i++;
            _context2.next = 55;
            break;

          case 80:
            if (!numImages) {
              _context2.next = 96;
              break;
            }

            DEBUG && log.debug('Completing [Content_Types].xml...');
            contentTypesPath = '[Content_Types].xml';
            _context2.next = 85;
            return (0, _zip.zipGetText)(zip, contentTypesPath);

          case 85:
            contentTypesXml = _context2.sent;
            _context2.next = 88;
            return (0, _xml.parseXml)(contentTypesXml);

          case 88:
            contentTypes = _context2.sent;

            // DEBUG && log.debug('Content types', { attach: contentTypes });
            ensureContentType = function ensureContentType(extension, contentType) {
              var children = contentTypes._children;
              if (children.filter(function (o) {
                return !o._fTextNode && o._attrs.Extension === extension;
              }).length) {
                return;
              }
              (0, _reportUtils.addChild)(contentTypes, (0, _reportUtils.newNonTextNode)('Default', {
                Extension: extension,
                ContentType: contentType
              }));
            };

            ensureContentType('png', 'image/png');
            ensureContentType('jpg', 'image/jpeg');
            ensureContentType('jpeg', 'image/jpeg');
            ensureContentType('gif', 'image/gif');
            finalContentTypesXml = (0, _xml.buildXml)(contentTypes, xmlOptions);

            (0, _zip.zipSetText)(zip, contentTypesPath, finalContentTypesXml);

          case 96:
            if (!replaceImages) {
              _context2.next = 118;
              break;
            }

            DEBUG && log.debug('Replacing images...');

            if (!options.replaceImagesBase64) {
              _context2.next = 117;
              break;
            }

            mediaPath = templatePath + '/media';
            imgNames = (0, _keys2.default)(replaceImages);
            _i = 0;

          case 102:
            if (!(_i < imgNames.length)) {
              _context2.next = 115;
              break;
            }

            imgName = imgNames[_i];
            imgPath = mediaPath + '/' + imgName;

            if ((0, _zip.zipExists)(zip, '' + imgPath)) {
              _context2.next = 108;
              break;
            }

            console.warn('Image ' + imgName + ' cannot be replaced: destination does not exist');
            return _context2.abrupt('continue', 112);

          case 108:
            imgData = replaceImages[imgName];

            DEBUG && log.debug('Replacing ' + imgName + ' with <base64 buffer>...');
            _context2.next = 112;
            return (0, _zip.zipSetBase64)(zip, imgPath, imgData);

          case 112:
            _i++;
            _context2.next = 102;
            break;

          case 115:
            _context2.next = 118;
            break;

          case 117:
            console.warn('Unsupported format (path): images can only be replaced in base64 mode');

          case 118:

            // ---------------------------------------------------------
            // Zip the results
            // ---------------------------------------------------------
            DEBUG && log.debug('Zipping...');
            _context2.next = 121;
            return (0, _zip.zipSave)(zip);

          case 121:
            output = _context2.sent;
            return _context2.abrupt('return', output);

          case 123:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, undefined);
  }));

  return function createReport(_x) {
    return _ref.apply(this, arguments);
  };
}();

// ==========================================
// Process images
// ==========================================
var processImages = function () {
  var _ref4 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3(images, documentComponent, zip, templatePath) {
    var imageIds, relsPath, relsXml, rels, i, imageId, _images$imageId, extension, imgData, imgName, imgPath, finalRelsXml;

    return _regenerator2.default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            DEBUG && log.debug('Processing images for ' + documentComponent + '...');
            imageIds = (0, _keys2.default)(images);

            if (!imageIds.length) {
              _context3.next = 38;
              break;
            }

            DEBUG && log.debug('Completing document.xml.rels...');
            relsPath = templatePath + '/_rels/' + documentComponent + '.rels';
            relsXml = void 0;
            _context3.prev = 6;
            _context3.next = 9;
            return (0, _zip.zipGetText)(zip, relsPath);

          case 9:
            relsXml = _context3.sent;
            _context3.next = 15;
            break;

          case 12:
            _context3.prev = 12;
            _context3.t0 = _context3['catch'](6);

            relsXml = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\n        <Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">\n        </Relationships>';

          case 15:
            _context3.next = 17;
            return (0, _xml.parseXml)(relsXml);

          case 17:
            rels = _context3.sent;
            i = 0;

          case 19:
            if (!(i < imageIds.length)) {
              _context3.next = 36;
              break;
            }

            imageId = imageIds[i];
            _images$imageId = images[imageId], extension = _images$imageId.extension, imgData = _images$imageId.data;
            imgName = 'template_image' + (i + 1) + extension;

            DEBUG && log.debug('Writing image ' + imageId + ' (' + imgName + ')...');
            imgPath = templatePath + '/media/' + imgName;

            if (!(typeof imgData === 'string')) {
              _context3.next = 30;
              break;
            }

            _context3.next = 28;
            return (0, _zip.zipSetBase64)(zip, imgPath, imgData);

          case 28:
            _context3.next = 32;
            break;

          case 30:
            _context3.next = 32;
            return (0, _zip.zipSetBinary)(zip, imgPath, imgData);

          case 32:
            (0, _reportUtils.addChild)(rels, (0, _reportUtils.newNonTextNode)('Relationship', {
              Id: imageId,
              Type: 'http://schemas.openxmlformats.org/officeDocument/2006/relationships/image',
              Target: 'media/' + imgName
            }));

          case 33:
            i++;
            _context3.next = 19;
            break;

          case 36:
            finalRelsXml = (0, _xml.buildXml)(rels, {
              literalXmlDelimiter: DEFAULT_LITERAL_XML_DELIMITER
            });

            (0, _zip.zipSetText)(zip, relsPath, finalRelsXml);

          case 38:
          case 'end':
            return _context3.stop();
        }
      }
    }, _callee3, undefined, [[6, 12]]);
  }));

  return function processImages(_x3, _x4, _x5, _x6) {
    return _ref4.apply(this, arguments);
  };
}();

// ==========================================
// Public API
// ==========================================
exports.default = createReport;
